Add support instructions to be displayed when creating an issue

marvin-roesch opened the pull request:
As of July 20, GitHub has added support (heh) for [support files](https://github.com/blog/2400-support-file-support). We appear to be getting reports that do not belong on this repo's issue tracker, like basic tech support or coding questions. In order to prevent this in the future, or to at least try and lower the amount of such reports, Forge should make use of such a support file.
My attempt at it is very detailed (hence the bold, short sentence at the top, for those who are really not up for reading something) and covers how to verify an issue and its source, how to determine the correct support channel and how to finally write the actual report. I tried to keep it as generic as possible for a modded Minecraft scenario and also give instructions for issues that don't have anything to do with Forge directly, since people might otherwise still report them here.
The instructions also include a basic issue template, which I can also add to the existing one if you want me to.

Any feedback on spelling or sentence structure would be appreciated. If there needs to be any clarification on a topic or you think some terms are too technical or complicated, please notify me as well. Keep in mind that the instructions are mainly aimed at players who have a minimal understanding of Java, if any at all, so I've put an emphasis on comprehensibility rather than technical correctness.

marvin-roesch committed:
Add support instructions to be displayed when creating an issue

mezz commented:
Looks good to me. There's already an issue template asking people to go to the forums but lots of people still ignore it. However, this will help people who actually want to write a good bug report, so why not?

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+ - **Retrace your last steps before the crash/bug occurred.** It might sound too simple, but a big part of reliably reproducing an issue is determining the steps required to trigger it. If you can't exactly remember what you last did, looking at the stacktrace again can also help you a bit, since there are often descriptive names in there which reflect an action you could take, such as using an i ...
+ - **Reduce the number of mod interactions.** You can achieve this in multiple ways. One of them is simply disabling any compatibility layers between involved mods via configuration files. The other one, which takes a little more effort, is actually removing mods from your instance. But don't be afraid, you won't have to remove mods one-by-one if you follow these steps (known as a binary search):
+   1. Move half the mods out of your instance (obviously keep those definitely participating in the issue).
+   2. If the issue persists after another try of reproducing it, repeat from step 1 one by halfing the mod list again.
```
- Noun: "half"
- Verb: "to halve"
    - Gerund: "halving"

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+### Preventing duplicates and unnecessary reports
+Use the target issue tracker's search functionality to look for issues which have descriptions similar to yours, but make sure to also search for already closed issues. If you can find one, verify it is in fact the same (or a very similar issue) and simply comment on it if you have anything else to contribute to the discussion, such as other steps to reproduce it. *Do not create duplicates of an  ...
+
+Additionally, always check the *latest* changelogs of Forge and participating mods to find out if the bug was already noticed and fixed by the developer. Often the maintainer may only support newer versions of the game, so also check the changelogs for versions of Minecraft you don't actually actively play. If the issue was adressed already, *do not report it, even if the fix only exists for a ne ...
```
> If the issue was adressed already, *do not report it, even if the fix only exists for a newer version of the game, since the developer probably had a good reason to do so.*

Placing (almost) a whole sentence in italics makes it sound forced. Maybe just italicize "do not report it", and either bold it or include the "even if ... game" part if you feel the need.

bs2609 replied:
How about "If the issue was addressed already, **do not report it**, _even if the fix only exists for a newer version of the game_, since the developer probably had a good reason to do so."?

howtonotwin replied:
Mixing emphasis like that feels icky to me, but it's really up to PC.

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+ - **Mod Version:** Just like the Forge and Minecraft version, this is crucial for quickly telling where an issue might come from. *If you're reporting an issue caused by Forge that only occurs with a certain mod installed, include this mod here. No mod version is required if it's a pure Forge bug.*
+ - **Complete list of mods:** In order to reliably analyze a bug, developers need the exact same context as you do when trying to find its cause. Hence, you should always include a *full list of mods with their corresponding version* in your report. Note that you should probably only include the ones you found by [isolating the issue](#isolating-an-issue).
+ - **Steps to reproduce:** You should always include a list of steps required to provoke an issue to occur. Without it, developers can only take rough guesses at what might cause the bug. Try to reduce the amount of steps to a bare minimum, to make reproducing the issue easier. If you require a sophisticated setup to get even close to the issue, you might want to consider including your save file ...
+ - **Full log:** This is probably the most important part of your bug report. You *always* have to include the full log (`fml-[client|server]-latest.log` from earlier), since it is the only way of telling what has caused an issue in code. *Do not* paste the full log as plain text in your report, but use a site like [Pastebin](https://pastebin.com) or [GitHub Gist](https://gist.github.com) instead ...
```
Casing in list is inconsistent

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+ 2. **Forge IRC Channel:** You can find a lot of help on the official [`#minecraftforge`](irc://irc.esper.net:6667/minecraftforge) channel on EsperNet. Especially smaller issues that you quickly need help with could be reported there. *Be aware that the channel is more suited towards development questions, so while you may get support there as a player, you should almost always prefer the forums. ...
+ 3. **Forge Issue Tracker:** If you're absolutely certain that you've found an issue in the Forge codebase (e.g. you're a modder yourself and have analyzed the stacktrace and relevant code), don't refrain from reporting it on this repository's [issue tracker](https://github.com/MinecraftForge/MinecraftForge/issues). *Please do not use the issues as a means of getting coding help. For developers,  ...
+
+### Mod Support
```
You used sentence case before, but are using title case here.

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+Getting Support surrounding Forge
```
It's fine if subheaders aren't title-cased, but this line should be, at least.

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+ 3. [Writing a helpful bug report](#writing-a-helpful-bug-report)
+    1. [Preventing duplicates and unnecessary reports](#preventing-duplicates-and-unnecessary-reports)
+    2. [Required Information](#required-information)
+    3. [Markdown Issue Template](#markdown-issue-template)
```
Casing consistency

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+
+You're most likely to get fast and valuable responses to your issues if you report them in the correct place. For this, you need to find the root cause of the problem.
+
+The most important source for this information are the log files Forge creates for every run of the game. You can find them in the `logs` subdirectory of your Minecraft game directory. The most important log file is `fml-[client|server]-latest.log` (depending on whether you run a singleplayer instance/connect to a server or are running a dedicated server) which contains all of Forge's and any mod ...
```
No point in asking readers to interpret `fml-[client|server]-latest.log`. List them out.

howtonotwin commented on SUPPORT.md:
````diff
@@ -0,0 +1,148 @@
+Exception in thread "..." path.to.Exception
+```
+
+This is followed by a number of `at` lines, which list the call chain that lead to the Exception being thrown. The further down you go this list, the earlier a call occurred. This is the most crucial part of any crash report, since it contains information about which mod or library caused an error.
````
This is pretty scary for a newcomer. Maybe: "Near the end of the log should be a section labelled 'Minecraft Crash Report', and after it, one about 'a detailed walkthrough of the error, ...'. In these sections, you'll see a bunch of lines that look like

```
    at some.words.related.to.a.mod.MoreWords(MoreWords.java:number)
```

"These lines are the most important part of a crash log, called the stacktrace, because they pinpoint the place in the code where something went irrecoverably wrong. Mod authors usually name their code so that these lines describe what the game was doing when it crashed. The further down one of these lines is, the earlier that piece of code ran, so at the bottom you'll see the place where Minecraft started, and at the top is where it stopped. You should use this information to figure out who to report the bug to."

marvin-roesch replied:
To be perfectly honest, I wrote this particular section without even looking at a crash report and that at a time when I wasn't supposed to be writing anything anymore, anyways. Thanks for your suggestion, I'll just use it :D

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+
+**Full log:** [See Paste]({link to gist/pastebin with full log})
+
+**Mods in this instance:**
```
Same inconsistency as above.

howtonotwin commented on SUPPORT.md:
```diff
@@ -0,0 +1,148 @@
+
+**Mod Version:** {mod version} // Remove this line if you're reporting to Forge and no specific mod is involved
+
+**Full log:** [See Paste]({link to gist/pastebin with full log})
```
Maybe just make the words "Full Log" a link.

marvin-roesch committed:
Use title case consistently for all headings/list items
Use @howtonotwin's paragraph introducing the stacktrace
Split client/server log to reduce ambiguity
